### PR TITLE
handle attachment storage in the service CORE-7595

### PR DESCRIPTION
**Pull request by mmaxim**
Patch does the following:

**Core**
1.) Moves most of the attachment code into a new package called `attachments`.
2.) Create a new class `AttachmentHTTPSrv` whose role is to create and serve URLs with attachment data in them. URLs are random and mapped to a `(conversation ID, msgID)` pair. The mapping is stored in memory in an LRU. `AttachmentHTTPSrv` takes a `AttachmentFetcher` parameter which runs how we get the actual attachment data for a request.
3.) Create `CachingAttachmentFetcher` which writes out the ciphertext of an attachment to disk, and serves plaintext to the frontend. Uses an on-disk LRU to keep the number of attachments stored to a sane level. 
4.) Change `DownloadFileAttachmentLocal` to use the current `AttachmentFetcher` so we can access our local cache.
5.) Change `PresentMessageUnboxed` to attach URLs for the preview and full attachment onto `UIMessage` for consumption by the frontend. This makes the flow for the frontend to read the URLs from these messages, and then immediately source the URLs. 

**Frontend**
1.) Remove a lot of Saga code involving creating files where attachments will be downloaded to.
2.) Pull the `devicePreviewPath` and `deviceFilePath` from the messages.
3.) Get rid of some `file://` hardcoded stuff.
4.) Use `DownloadFileAttachmentLocal` for doing an actual download, instead of doing a copy from the local cache.
5.) Touch some other random stuff up (like processing `ATTACHMENTUPLOADED` messages differently).

**TODO**

- [ ] Loading state of an image no longer has a gray box (it's just blank).
- [ ] Share an attachment no longer works on mobile. Will need some system to download the file to a temporary to share (and remove the file).
- [ ] Error state of an image is now a broken image on desktop (like if it is being uploaded by the sender still). Probably should sub in either the gray box (which doesn't sound great), or a spinner if possible. 

@keybase/react-hackers I'm handing this over to you to finish this up. It would be great if it could get merged before I leave for vacation so I can use it!

Also cc @maxtaco if you are interested.

### CODE REVIEW

File: go/chat/attachments/utils.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/chat1"
+	"github.com/keybase/client/go/protocol/gregor1"
+	"golang.org/x/net/context"
+)
+
+func AssetFromMessage(ctx context.Context, g *globals.Context, uid gregor1.UID, convID chat1.ConversationID,
+	msgID chat1.MessageID, preview bool) (res chat1.Asset, err error) {
+
+	msgs, err := g.ChatHelper.GetMessages(ctx, uid, convID, []chat1.MessageID{msgID}, true)
```
Comment (patrickxb) [inline go/chat/attachments/utils.go]: missing err check

File: shared/actions/chat2/index.js [hunk 2]
```diff
@@ -1538,20 +1438,11 @@ function* attachmentLoad(action: Chat2Gen.AttachmentLoadPayload) {
...
-    const result = yield Saga.call(downloadFileRpc.run)
-    if (EngineRpc.isFinished(result)) {
-      yield Saga.put(Chat2Gen.createAttachmentLoaded({conversationIDKey, isPreview, ordinal, path: fileName}))
-    } else {
-      yield Saga.put(Chat2Gen.createAttachmentLoadedError({conversationIDKey, isPreview, ordinal}))
-    }
-  } catch (err) {
-    logger.warn('attachment failed to load:', err)
-    yield Saga.put(Chat2Gen.createAttachmentLoadedError({conversationIDKey, isPreview, ordinal}))
+  const result = yield Saga.call(downloadFileRpc.run)
```
Comment (chrisnojima) [inline shared/actions/chat2/index.js]: i'm worried about ditching the try block and err catching. maybe we don't handle this nicely w/ the downloads

File: shared/actions/json/chat2.json [hunk 3]
```diff
@@ -67,7 +67,7 @@
         "'inboxSyncedUnknown'",
         "'joinedAConversation'",
         "'leftAConversation'",
-        "'teamTypeChanged'",
```
Comment (chrisnojima) [inline shared/actions/json/chat2.json]: i guess your linter doens't like dangling commas?

File: shared/desktop/renderer/renderer.html [hunk 4]
```diff
@@ -1,14 +1,18 @@
...
-    <div id="root">
-      <div title="loading..." style="flex: 1;background-color: #f5f5f5"></div>
-    </div>
-    <div id="modal-root"></div>
-    <script src="renderer-load.js"></script>
-  </body>
-</html>
+
+<head>
+  <meta charset="utf-8" http-equiv="Content-Security-Policy" content="default-src file: filesystem: http://localhost:4000 https://keybase.io/; img-src data: http://localhost:4000 http://127.0.0.1:* https://keybase.io/ https://pbs.twimg.com/ https://avatars.githubusercontent.com/ https://images.coinbase.com/ https://s3.amazonaws.com/keybase_processed_uploads/ file:; style-src 'unsafe-inline'; script-src file: http://localhost:4000 chrome-extension://react-developer-tools 'unsafe-eval'; connect-src ws://localhost:4000 http://localhost:4000"
```
Comment (chrisnojima) [inline shared/desktop/renderer/renderer.html]: local isn't https? does it matter?

File: shared/actions/chat2/index.js (hunk 2 above)
Comment (mmaxim) [reply shared/actions/chat2/index.js]: I can put back the loaded error action I suppose. I'm not sure what it did though.

File: shared/actions/json/chat2.json (hunk 3 above)
Comment (mmaxim) [reply shared/actions/json/chat2.json]: This file is just auto generated right?

File: shared/desktop/renderer/renderer.html (hunk 4 above)
Comment (mmaxim) [reply shared/desktop/renderer/renderer.html]: Don't think it matters, would need root to get any traffic there.

File: shared/actions/json/chat2.json (hunk 3 above)
Comment (chrisnojima) [reply shared/actions/json/chat2.json]: opposite, this is the source of the generation. its fine though, just mentioning it

Comment (mmaxim) [reply shared/actions/json/chat2.json]: Ah right, I guess it did make a bunch of changes.

File: go/chat/server.go [hunk 5]
```diff
@@ -1675,6 +1651,10 @@ func (h *Server) DownloadAttachmentLocal(ctx context.Context, arg chat1.Download
 	ctx = Context(ctx, h.G(), arg.IdentifyBehavior, &identBreaks, h.identNotifier)
 	defer h.Trace(ctx, func() error { return err }, "DownloadAttachmentLocal")()
 	defer func() { err = h.handleOfflineError(ctx, err, &res) }()
+	if err = h.assertLoggedIn(ctx); err != nil {
```
Comment (joshblum) [inline go/chat/server.go]: why not `h.assertLoggedInUID`?

Comment (mmaxim) [reply go/chat/server.go]: copy/pasted

Comment (mmaxim) [reply go/chat/server.go]: Actually that code has been there forever, if I changed it wasn't anything significant.

Comment (joshblum) [reply go/chat/server.go]: oh ok i just noticed we never use `assertLoggedInUID` throughout the file even when we also need a uid and was wondering why

**Pull request merged**